packages/library/src/base/plugin.ts

```diff
@@ -1,13 +1,23 @@
 import { without } from 'lodash'
 import { Component } from './component'
+import { getEventMethodName } from './util/eventName'
 
 type PluginEvent = 'plugin:add' | 'plugin:remove'
 
-export class Plugin<C extends Component = Component, E = string> {
-  async handle(context: C, event: E | PluginEvent, data?: any) {}
+export class Plugin<
+  C extends Component = Component,
+  E extends string = string,
+> {
+  async handle(context: C, event: E | PluginEvent, ...params: any[]) {
```

Comment:
The dynamic dispatching implementation here is clever, I think there's a good way to do it in TS!

First off, it might be easier just to hardcode a handful of empty methods in this base class here and manually trigger them from the handle.

However, if you want to go full TypeScript wizard, it is now possible with [Template Literal Types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) to perform string manipulation on keys within types. To say, define a number of `on${Event}` methods computed from a union type of events. Here's a tsplayground where I play around with it: https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbhAdsAcgQwLbQLxQOQBOAronlAD75gERho1mV5IAmeAUO6JLAsgBJpELADYQCAYRFoAzjKi4A3uyiqoAbQDWEEFACWiKAAMA9ogAkiiWjB7gaEXoBeEADzwkqTBAB8AXyMAXQAuKAAKAEoFHyg4Ez0Wdj9OAGNpOV5PAAURIgBzA1cJKAgAD2BWeQAjExMxIQAaWBLyyszkdCwYvQwwMSxkeQ8BIVFxKVl5ZTUoMwAlEkioZRU1ZJmzLJo6GiWVmfW1MxhhPc4DpKggA

Reply:
Whoa, thanks! I had no idea this was possible in TS, this is truly excellent. I'll see if I can incorporate your example, many thanks for that!

Given the limited number of intrinsic string manipulation types it probably makes sense to rename some events, specifically `plugin:add` to `pluginAdd` etc., but that strikes me as a good idea anyway; as far as I know that's the only place we use the good ol' [`backbone.js` event syntax](http://backbonejs.org/#Events-catalog).